### PR TITLE
only run --all-targets in stage0 for Std

### PR DESCRIPTION
**Pull request by jo3bingham**
Repro'd the issue with `python3 x.py check --stage 1 library/std` and tested the fix with the same command.

r? @jyn514 

I tried to implement [this solution](https://github.com/rust-lang/rust/issues/98720#issuecomment-1184435462), but didn't have any luck. I don't think I fully understood what needed to be done. However, I would love to be mentored on it since it would be a more correct solution, and I can learn more about how bootstrap works.

### CODE REVIEW

**Comment by jyn514** [comment]
> I tried to implement this solution, but didn't have any luck.

When you say didn't have luck, you mean it didn't fix the problem? Or you didn't understand it well enough to write code for it?

I think the code should be fairly simple, basically copy those two lines immediately before `--all-targets`.

File: src/bootstrap/check.rs [hunk 1]
```diff
@@ -140,7 +140,10 @@ impl Step for Std {
             cargo_subcommand(builder.kind),
         );
 
-        cargo.arg("--all-targets");
+        if compiler.stage == 0 {
```
Comment (jyn514) [inline src/bootstrap/check.rs:143]: Please add a comment here; adapting my explanation on the issue is fine, but it's very unclear right now why stage 0 is special cased.

**Comment by jo3bingham** [comment]
> When you say didn't have luck, you mean it didn't fix the problem? Or you didn't understand it well enough to write code for it?
> 
> I think the code should be fairly simple, basically copy those two lines immediately before `--all-targets`.

I wasn't sure if it was because I didn't understand it well enough or just that it didn't fix the problem, but I did copy those two lines before `--all-targets` and the errors were the same for me as it was without it:
```rust
builder.ensure(crate::compile::Std::new(compiler, compiler.host));
builder.ensure(crate::compile::Std::new(compiler, target));
cargo.arg("--all-targets");
```
```
error: could not compile `getrandom` due to 125 previous errors
error: could not compile `test` due to 566 previous errors
error: could not compile `proc_macro` due to 552 previous errors
```

**Comment by jyn514** [comment]
> I was mistaken, the errors I get from adding those two lines are different. Sample error and logs:

Is this with or without calling `add_to_sysroot`? you can only do one or the other, calling add_to_sysroot *and* `ensure(compile::Std)` will lead to errors.

**Comment by jyn514** [comment]
:+1: this seems good to me - it works around the error, and it seems exceedingly rare to want to check std tests compile without actually running them (in stage1, no less).

@bors r+ rollup

**Pull request merged**